### ISSUE TITLE
Wrong sniff codes with "special namespace"

### ISSUE THREAD

**Issue by rhorber** [squizlabs/PHP_CodeSniffer#1569]
When using a namespace for sniffs which ends not with `StandardName\Sniffs`, the sniff codes in reports are wrong (the  `Standard.*` ones are these of my own sniffs):
`[ ] Generic.PHP.NoSilencedErrors.Discouraged                      3`
`[ ] Standard.Commenting.FunctionComment.SpacingAfterType          3`
`[ ] Standard.Commenting.Todo.TaskFound                            2`
`[ ] Standard.Strings.Label.NoLabel                                2`

IMHO this should be changed. 
I created [PR 1568](https://github.com/squizlabs/PHP_CodeSniffer/pull/1568) to fix this.

**Comment by rhorber** [comment]
**TL;DR: The PR above leads to inconsistencies.**

I found another issue (bug?) where the same behaviour occurres: When passing sniff codes as `sniffs` or ` exclude` argument.

Here's the beginning of my ruleset file:
`<?xml version="1.0"?>`
` <ruleset name="Rhorber Intranet" namespace="Rhorber\PHPCS\Standard">`
` <description>`

The example below is with the `autoload.php` file from PHP_CodeSniffer. (Its pointless checking it against my rules, but explains the problem very well.)

When calling ` phpcs --standard=Rhorber autoload.php -s` I get this:
[autoload-all.txt](https://github.com/squizlabs/PHP_CodeSniffer/files/1166977/autoload-all.txt)

Limiting to a sniff code from my standard (`phpcs --standard=Rhorber autoload.php -s --report-file=autoload-all --sniffs=Rhorber.Commenting.CommentStructure`) gets me the erorr `ERROR: No sniffs were registered`.

The sniff code above (`Rhorber.`) is returned with the changes of my PR above.
Limiting with the sniff code from the report (without my changes) leads to results (`phpcs --standard=Rhorber autoload.php -s --report-file=autoload-sniffs.txt --sniffs=Standard.Commenting.CommentStructure`):
[autoload-sniffs.txt](https://github.com/squizlabs/PHP_CodeSniffer/files/1166981/autoload-sniffs.txt)

The `--exclude` switch does not exclude any errors/warning when passed codes with `Rhorber.*`. But it works correctly with `Standard.*` (in this scenario).

The behaviour as it is, is consistent. The codes in the reports are the ones that can be passed and everything works fine.
If the codes should be printed correctly, when using a non-default namespace, my PR does not the whole job...

So my question is, what to do next?
- Are these "wrong" sniff codes no real bugs, and my PR is obsolete?
- Should the printed codes be fixed, but my PR is comletely wrong?
- Should I (or somebody else) extend the PR by adapting the behaviour of `sniffs` and `extends`

I'm fine with all of them. Let me know what's your opinion and where I can help :-)

**Comment by rhorber** [comment]
It bugged me, so I tried to change the code, that `--sniffs` and `--extends` work with my setup.
I updated the PR ([1568](https://github.com/squizlabs/PHP_CodeSniffer/pull/1568)).

As mentioned above, if the changes are wrong or my scenario is not intended, feel free to delete it :)

**Comment by gsherwood** [comment]
There is currently an absolute requirement that sniff namespaces end with the format StandardName/Sniffs/CategoryName. It's documented here: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Version-3.0-Upgrade-Guide#upgrading-custom-sniffs

If this format doesn't exist, PHPCS can't know what the standard name code is just by looking at the namespace. As in your PR, you could try and source it from the plain text name provided in a ruleset.xml file, but this assumes that the name is exactly the same as a value in the namespace, which is not always the case. I know, for example, that the WordPress coding standards use spaces in the standard name so they print nicely when you use the `-i` CLI arg. The PR would break that standard.

So I think it is best to leave the rules as they are to ensure custom coding standards keep working.

A possible solution would be to add another attribute to the ruleset tag that allows you to specify the standard's code name if you've used a custom namespace format that does not match the PHPCS format. I'm not sure if I'd do this the way you've structured your PR (I'd need to think about it a bit more) but I think it is achievable and would allow you to keep your namespace and still get rid of that `Standard` name in the sniff codes.

I'll add this to a release milestone so I can take a look at it later on.

**Comment by rhorber** [comment]
My bad, sorry. I did not read it carefully enough (my situation is explicitly noted), I got misled by the second note.

I did not realise, that the standard name has to be in the namespace in my solution, which is not ideal...

I agree with you, to keep everything as is.

In that case, I just live with it or change the namespace of my ruleset to the rules.

Thanks.

**Comment by gsherwood** [comment]
3.2.0 is looking pretty full and this hasn't come up again, so I'm going to remove it from the release but leave this as an open feature in case it becomes more widely useful.

**Comment by gsherwood** [comment]
Closing as this hasn't received any additional support.
